akka-actor-typed/src/main/scala/akka/actor/typed/delivery/internal/ConsumerControllerImpl.scala

```diff
@@ -0,0 +1,467 @@
+    context.log.info("from producer [{}], deliver [{}] to consumer", seqMsg.producerId, seqMsg.seqNr)
+    s.consumer ! Delivery(seqMsg.producerId, seqMsg.seqNr, seqMsg.msg, context.self)
+    waitingForConfirmation(s, seqMsg)
+  }
+
+  // The message has been delivered to the consumer and it is now waiting for Confirmed from
+  // the consumer. New SequencedMessage from the ProducerController will be stashed.
+  private def waitingForConfirmation(s: State[A], seqMsg: SequencedMessage[A]): Behavior[InternalCommand] = {
+    Behaviors.receiveMessage {
+      case Confirmed(seqNr) =>
+        val expectedSeqNr = s.receivedSeqNr
+        if (seqNr > expectedSeqNr) {
+          throw new IllegalStateException(
+            s"Expected confirmation of seqNr [$expectedSeqNr], but received higher [$seqNr]")
+        } else if (seqNr != expectedSeqNr) {
+          context.log.info(
+            "Expected confirmation of seqNr [{}] but received [{}]. Perhaps the consumer was restarted.",
+            expectedSeqNr,
+            seqNr)
+        }
+        context.log.info("Confirmed [{}], stashed size [{}]", seqNr, stashBuffer.size)
+
+        val newRequestedSeqNr =
+          if (seqMsg.first) {
+            // confirm the first message immediately to cancel resending of first
+            val newRequestedSeqNr = seqNr - 1 + flowControlWindow
+            context.log.info("Request after first [{}]", newRequestedSeqNr)
+            s.producer ! Request(confirmedSeqNr = seqNr, newRequestedSeqNr, resendLost, viaTimeout = false)
+            newRequestedSeqNr
+          } else if ((s.requestedSeqNr - seqNr) == flowControlWindow / 2) {
+            val newRequestedSeqNr = s.requestedSeqNr + flowControlWindow / 2
+            context.log.info("Request [{}]", newRequestedSeqNr)
+            s.producer ! Request(confirmedSeqNr = seqNr, newRequestedSeqNr, resendLost, viaTimeout = false)
+            startRetryTimer() // reset interval since Request was just sent
+            newRequestedSeqNr
+          } else {
+            if (seqMsg.ack) {
+              context.log.info("Ack [{}]", seqNr)
+              s.producer ! Ack(confirmedSeqNr = seqNr)
+            }
+
+            s.requestedSeqNr
+          }
+
+        // FIXME can we use unstashOne instead of all?
+        if (stashBuffer.nonEmpty)
+          context.log.info("Unstash [{}]", stashBuffer.size)
+        stashBuffer.unstashAll(active(s.copy(confirmedSeqNr = seqNr, requestedSeqNr = newRequestedSeqNr)))
+
+      case seqMsg: SequencedMessage[A] =>
+        if (stashBuffer.isFull) {
+          // possible that the stash is full if ProducerController resends unconfirmed (duplicates)
+          // dropping them since they can be resent
+          context.log.info("Stash is full, dropping [{}]", seqMsg)
+        } else {
+          context.log.info("Stash [{}]", seqMsg)
+          stashBuffer.stash(seqMsg)
+        }
+        Behaviors.same
+
+      case Retry =>
```

Comment:
The retrying can be quite noisy if the producer is slow. Could we change it so that the timer is reset when we receive a SequencedMessage (even if it is stashed?) and perhaps make it an exponential backoff to quite a low max for when things are just quiet?

Reply:
made a TODO of that for now. I agree that it would be nice to optimize that I think it has to be more sophisticated than just rescheduling for each incoming